File: SuperheroApp/Characters/Domain/UseCase/GetCharactersUseCase.swift
//
//  CharactersUseCase.swift
//  SuperheroApp
//
//  Created by Balazs Varga on 2018. 02. 12..
//  Copyright © 2018. W.UP. All rights reserved.
//

struct GetCharactersRequest: UseCaseRequest {
    var page: Page
    
    init(page: Page) {
        self.page = page
    }
}

struct GetCharactersResponse: UseCaseResponse {
    var characters: [Character]
}

class GetCharactersUseCase: UseCase<GetCharactersRequest, GetCharactersResponse> {
    
    private let charactersDataSource: CharactersDataSource
    
    init(charactersDataSource: CharactersDataSource) {
        self.charactersDataSource = charactersDataSource
    }
    
    override func executeUseCase(request: GetCharactersRequest) throws {
        self.charactersDataSource.loadCharacters(page: request.page, complete: { (characters: [Character]) in
            let response = GetCharactersResponse(characters: characters)
            if let success = self.success {
                success(response)
            }
        }, fail: {
            if let error = self.error {
                error()
            }
        })
    }
}

File: common/src/main/kotlin/digital/wup/superheroapp/characters/domain/usecase/GetCharactersUseCase.kt
package digital.wup.superheroapp.characters.domain.usecase

import digital.wup.superheroapp.UseCase
import digital.wup.superheroapp.UseCaseRequest
import digital.wup.superheroapp.UseCaseResponse
import digital.wup.superheroapp.characters.domain.usecase.model.Character
import digital.wup.superheroapp.characters.domain.usecase.model.Page
import digital.wup.superheroapp.datasource.CharactersDataSource

class GetCharactersRequest(val page: Page) : UseCaseRequest

class GetCharactersResponse(val characters: List<Character>): UseCaseResponse

class GetCharactersUseCase(private val charactersDataSource: CharactersDataSource): UseCase<GetCharactersRequest, GetCharactersResponse>() {

    override fun executeUseCase(request: GetCharactersRequest) {
        charactersDataSource.loadCharacters(request.page, { characters: List<Character> ->
            val response = GetCharactersResponse(characters)
            success(response)
        }, {
            error()
        })
    }
}